psc-publish/Main.hs

```diff
@@ -196,18 +197,35 @@ getResolvedDependencies :: [PackageName] -> PrepareM [(PackageName, Version)]
 getResolvedDependencies declaredDeps = do
   depsBS <- fromString <$> readProcess' "bower" ["list", "--json", "--offline"] ""
 
-  deps <- catchLeft (parse asBowerResolvedDependencies depsBS)
-                    (internalError . JSONError FromBowerList)
+  -- Check for undeclared dependencies
+  toplevels <- catchJSON (parse asToplevelDependencies depsBS)
+  warnUndeclared declaredDeps toplevels
 
-  warnUndeclared declaredDeps (map fst deps)
+  deps <- catchJSON (parse asResolvedDependencies depsBS)
   handleDeps deps
 
   where
-  asBowerResolvedDependencies ::
-    Parse BowerError [(PackageName, DependencyStatus)]
-  asBowerResolvedDependencies =
-    key "dependencies"
-      (eachInObjectWithKey (parsePackageName . T.unpack) asDependencyStatus)
+  catchJSON = flip catchLeft (internalError . JSONError FromBowerList)
+
+-- | Extracts all dependencies and their versions from
+--   `bower list --json --offline`
+asResolvedDependencies :: Parse BowerError [(PackageName, DependencyStatus)]
+asResolvedDependencies = nubBy ((==) `on` fst) <$> go
+  where
+  go =
+    fmap (fromMaybe []) $
+      keyMay "dependencies" $
+        (++) <$> eachInObjectWithKey (parsePackageName . T.unpack)
+                                     asDependencyStatus
+             <*> (concatMap snd <$> eachInObject asResolvedDependencies)
```

Comment:
If there is a cyclic dependency (unlikely but possible, I think), won't this `<<loop>>`?